### PR TITLE
Fix and add unit tests for Caps Word to work with Unicode Map, Auto Shift, Retro Shift.

### PR DESCRIPTION
**Pull request by getreuer**
This PR fixes Caps Word to work correctly with Unicode Map, Auto Shift, and Retro Shift, and adds unit tests.

## Description

Caps Word currently does not work correctly with Unicode Map, Auto Shift, and Retro Shift:

* With Unicode Map pairs `XP(a, A)`, it should be possible to handle these keycodes in `caps_word_press_user()` such that they are shifted. But currently, Unicode Map's handler runs before Caps Word, and Caps Word does not see the event.

* When Auto Shift is enabled for letter keys, they should be shifted when Caps Word is on regardless of whether they are released before `AUTO_SHIFT_TIMEOUT`, but currently they are not. Similarly for Retro Shift.

This PR is:

1. @kosorin's https://github.com/qmk/qmk_firmware/pull/17247, which fixes Caps Word to work with Unicode Map. Incidentally, it also fixes Caps Word to work with Auto Shift and Retro Shift, which currently do not work correctly as @qyurila noted in https://github.com/qmk/qmk_firmware/pull/17240.

2. Unit tests to verify that these features now work together. 

## Types of Changes

- [x] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by drashna — APPROVED** [review]

File: tests/caps_word/caps_word_autoshift/test_caps_word_autoshift.cpp
Comment (KarlK90) [inline tests/caps_word/caps_word_autoshift/test_caps_word_autoshift.cpp:42]: How about moving these helper methods into the base `TestFixture` class? I have seen it added in other unit-tests as well and future unit-tests would surely profit from them as well.

File: tests/caps_word/caps_word_unicodemap/test_caps_word_unicodemap.cpp
Comment (KarlK90) [inline tests/caps_word/caps_word_unicodemap/test_caps_word_unicodemap.cpp:86]: Can this function be used universally by other tests that exercise Unicode emulation? If so this should become part of `TestDriver` and probably (?) be written in all caps to follow the Gtest convention on assertion macros.

**Review by KarlK90 — CHANGES_REQUESTED** [review]
Thanks for you contribution and also that you came back and added more unit-tests! Really appreciated. I have some recommendations on moving the convenience test helpers into base classes to make them universally accessible for other unit-test but apart from that you have my approval.

File: tests/caps_word/caps_word_autoshift/test_caps_word_autoshift.cpp
Comment (getreuer) [reply tests/caps_word/caps_word_autoshift/test_caps_word_autoshift.cpp]: Done. Added `tap_key(key, delay_ms)` and `tap_keys(key_a, key_b, ...)` methods to TestFixture.

File: tests/caps_word/caps_word_unicodemap/test_caps_word_unicodemap.cpp
Comment (getreuer) [reply tests/caps_word/caps_word_unicodemap/test_caps_word_unicodemap.cpp]: Done. Added an `EXPECT_UNICODE` macro that sets a gmock expectation that a Unicode code point will be sent.

While I was at it, I also added an `EXPECT_REPORT` macro for a shorthand notation to expect a given keyboard report, since this is so frequent in the tests.

**Comment by getreuer** [comment]
@KarlK90 thank you for the suggestions! I added some new test utilities to this PR: `EXPECT_REPORT`, `EXPECT_UNICODE`, `tap_key()`, and `tap_keys()`. (Let me know if you had a different direction in mind.)

I went ahead and applied them to most tests. Particularly the `EXPECT_REPORT` macro is frequently useful and makes a nice reduction in boilerplate in many tests. 

Here is a description of the utilities (also documented in the code):

* `EXPECT_REPORT` sets a gmock expectation that a given keyboard report will be sent. For instance,

  ```c
  EXPECT_REPORT(driver, (KC_LSFT, KC_A));
  ```

  is shorthand for

  ```c
  EXPECT_CALL(driver,
    send_keyboard_mock(KeyboardReport(KC_LSFT, KC_A)));
  ```

* `EXPECT_UNICODE` sets a gmock expectation that a given Unicode code point will be sent using UC_LNX input mode. For instance for U+2013,

  ```
  EXPECT_UNICODE(driver, 0x2013); 
  ```

  expects the sequence of keys: "Ctrl+Shift+U, 2, 0, 1, 3, space".

* `tap_key(key)` is a TestFixture method that taps a given KeymapKey, optionally with a specified delay between press and release.

* Similarly, the method `tap_keys(key_a, key_b, key_c)` taps a sequence of KeymapKeys.

**Comment by KarlK90** [comment]
@getreuer Those facilities and changes to the tests are very nice indeed! Improving the ergonomics for writing these tests is one of my long term goals and this is a step in the right direction. 

While reading the changes to the unit tests I noticed that we have two more frequent use-cases that we expect:

* empty reports `EXPECT_REPORT(driver, ());`
* making sure that no reports are send `EXPECT_CALL(driver, send_keyboard_mock(_)).Times(0);`

These can easily be added by providing these macros as well:

```
#define EXPECT_EMPTY_REPORT(driver) EXPECT_REPORT(driver, ())

#define EXPECT_ANY_REPORT(driver) EXPECT_CALL((driver), send_keyboard_mock(_))

#define EXPECT_NO_REPORT(driver) EXPECT_ANY_REPORT(driver).Times(0)
```

I have them in a working condition in my local branch, it juts needs some clean-up and documentation. If you like/don't mind I can go ahead and push them to this PR.

**Comment by getreuer** [comment]
@KarlK90 Yes, please go ahead and push into this PR! =)

Those are great additions. Not only is it more concise, but the meaning of `EXPECT_ANY_REPORT(driver)`  is more readable and obvious than

```cpp
EXPECT_CALL(driver, send_keyboard_mock(_));
```

**Comment by KarlK90** [comment]
@getreuer Done! I amended the changes to you commits and changed the commit messages. I took the liberty to convert the documentation comments to doxygen, but left them untouched otherwise. Expect for the empty report example in `EXPECT_REPORT`.

**Review by KarlK90 — APPROVED** [review]

**Comment by getreuer** [comment]
@KarlK90  Awesome! This all looks good. It is fine by me to merge this PR whenever the maintainers would like to do so.

**Comment by KarlK90** [comment]
Now also on the develop branch :tada:

**Pull request merged**
